src/traces/carpet/calc_gridlines.js

```diff
@@ -0,0 +1,340 @@
+/**
+* Copyright 2012-2017, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var Axes = require('../../plots/cartesian/axes');
+var extendFlat = require('../../lib/extend').extendFlat;
+
+module.exports = function calcGridlines(trace, axisLetter, crossAxisLetter) {
+    var i, j, j0;
+    var eps, bounds, n1, n2, n, value, v;
+    var j1, v0, v1, d;
+
+    var data = trace[axisLetter];
+    var axis = trace[axisLetter + 'axis'];
+
+    var gridlines = axis._gridlines = [];
+    var minorgridlines = axis._minorgridlines = [];
+    var boundarylines = axis._boundarylines = [];
+
+    var crossData = trace[crossAxisLetter];
+    var crossAxis = trace[crossAxisLetter + 'axis'];
+
+    if(axis.tickmode === 'array') {
+        axis.tickvals = [];
+        for(i = 0; i < data.length; i++) {
+            axis.tickvals.push(data[i]);
+        }
+    }
+
+    var xcp = trace._xctrl;
+    var ycp = trace._yctrl;
+    var nea = xcp[0].length;
+    var neb = xcp.length;
+    var na = trace.a.length;
+    var nb = trace.b.length;
+
+    Axes.calcTicks(axis);
```

Comment:
somehow we'd better get `layout.separators` in there instead of hardcoding the defaults...